WIP: Resurrection

NotBad4U opened the pull request:
# Resurrection process

![](https://vignette.wikia.nocookie.net/wowwiki/images/1/1c/RedemptionTCG.jpg/revision/latest?cb=20121220030542)

I'm close to finish the resurrection of this great project. Where I still don't understand how I was invoked to in. We have some stuff to do (listed below) to finish the resurrection process.

## Remaining work

- [ ]  Fix the test request_message_too_long_tests
- [ ]  Fix the test message_set_too_long_tests
- [x]  Fix the test kafka_array_test
- [ ]  Fix the test request_message_too_short_tests
- [ ]  Test it with a `kafka client`

NotBad4U committed:
Upgrade dependencies version

NotBad4U committed:
Use copy_from_slice instead of copy_memory which is deprecated

NotBad4U committed:
Refactore Handler with the last mio version

NotBad4U committed:
Add primitiv parser

One test fail take care

NotBad4U committed:
Fix offset parser

Switch chain! to do_parse

NotBad4U committed:
Fix metadata parser

NotBad4U committed:
Fix offset commit parser

NotBad4U committed:
Fix offset_fetch parser

NotBad4U committed:
Fix consumer metadata parser

NotBad4U committed:
Fix fetch parser

NotBad4U committed:
Fix compilation warnings

NotBad4U committed:
Fix message, procude and request parser

NotBad4U committed:
Fix request parser

Take care some tests failed

NotBad4U committed:
Fix response primitive

juste removed use nom::Err::*;

NotBad4U committed:
Add reponse metadata

NotBad4U committed:
Fix response produce

NotBad4U committed:
Fix response consumer_metadata

NotBad4U committed:
Fix response fecth

NotBad4U committed:
Fix reponses offets

Just removed the nom::Err

NotBad4U committed:
Fix response

Just removed the nom::Err

NotBad4U committed:
Fix handle request method

NotBad4U committed:
Fix the handling message

Migrate the read_slice which no longer exist to put_slice. Should be the same effect

NotBad4U committed:
Fix kafka_array_test, nom 3.0 seem to have change the behavior of be_i8 which call the result of map on be_u8. Actually be_u8 return Incomplete(Needed::Size(1)) instead of Incomplete(Needed::Unknown)

Geal commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Proust. A stream of fresh air
+# Proust
```
but why remove that? :'(

NotBad4U replied:
Because I hate this kind of humor :stuck_out_tongue_closed_eyes:  srry

Geal commented on src/main.rs:
```diff
@@ -17,21 +17,13 @@ mod responses;
-
 fn main() {
-  //storage::storage_test();
-  println!("Le peintre original procède à la façon des oculistes.");
```
why remove that too? :'(

NotBad4U replied:
Okai I re added this in cf158e6

Geal commented on src/main.rs:
```diff
@@ -17,21 +17,13 @@ mod responses;
-}
+  storage::storage_test();
 
+  let zjg = network::zookeeper::start_listener("127.0.0.1:2088".to_string()).expect("start zookeeper");
```
we won't need the zookeeper part, so no need to start the server for this

NotBad4U replied:
Removed

Geal commented on src/network/handler.rs:
```diff
@@ -1,89 +1,104 @@
 
   fn read_size(&mut self) -> ClientResult {
-    let mut size_buf = ByteBuf::mut_with_capacity(4);
+    let mut size_buf = BytesMut::with_capacity(4);
```
we don't need a BytesMut for 4 bytes. a `[u8; 4]`on the stack should be enough

NotBad4U replied:
Yeah thks I modify this in the commit 40f4dee

Geal commented on src/network/handler.rs:
```diff
@@ -1,89 +1,104 @@
-          //println!("found size: {}", sz);
+        }
+        else {
+          // Please forgive me...
```
if you need to parse an integer from the network, there are methods available for this in nom

NotBad4U replied:
Yeah thks I modify this in the commit 40f4dee

Geal commented on src/network/handler.rs:
```diff
@@ -101,24 +116,21 @@ pub trait NetworkClient {
-      },
-      Ok(None) => Err(ClientErr::Continue),
+    match self.socket().write(msg) {
+      Ok(size)  => Ok(size),
```
maybe this should be done in a loop (because `size` might be smaller than `msg.len()`)

NotBad4U replied:
done in 65c51b2
I use a write_all instead of a loop

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
+pub struct Server<C: Client> {
+  pub tcp_listener: TcpListener,
+  pub token_index:  usize,
+  pub clients:      HashMap<usize, C>,
```
maybe use a `Slab` from the slab crate here, it will be nicer to use

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
+
+  pub fn new(addr: SocketAddr, poll: Poll) -> Self {
+    Server {
+      tcp_listener: TcpListener::bind(&addr.into()).unwrap(),
```
return an error instead of unwrapping here. Not being able to start a listener is common (maybe the IP is wrong, or someone already uses the port)

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
+      token_index: 1,
+      clients: HashMap::new(),
+      poll,
+      available_tokens: Vec::new()
```
we won't need availabltokens if we use a slab

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
+          Token(t) => {
+            let kind = event.readiness();
+
+            if UnixReady::from(kind).is_hup() {
```
please use a serie of `if ... {` instead of if/else, because the event's kind might contain multiple things (like, the socket is both readable and hup). It might be good to store the previous readiness for the socket, too. Like, maybe we weren't able to read the whole buffer, so we stop, but the socket is still readable

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
+            else if UnixReady::from(kind).is_readable() {
+              self.client_read(t);
+            }
+            else {
```
with the current code, if kind is an error, we will try to write instead of handling the error

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
-  fn client_read(&mut self, event_loop: &mut EventLoop<Self>, tk: usize) {
-    //println!("client n°{:?} readable", tk);
-    if let Some(mut client) = self.clients.get_mut(&tk) {
+  fn next_token(&mut self) -> usize {
```
not needed with a slab

Geal commented on src/network/handler.rs:
```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
-
-              let capacity = buffer.remaining();  // actual buffer capacity may be higher
-              //println!("capacity: {}", capacity);
+              let mut buffer = BytesMut::with_capacity(size);
```
it's probably a bad idea to allocate a `BytesMut` each time we want to read. I know the previous code did that, but it's not really a good thing. Maybe we'll put that in an issue to fix it later

Geal commented on src/network/kafka.rs:
```diff
@@ -43,17 +32,18 @@ impl NetworkClient for Client {
+
+    buffer.put_slice(&mut res[..]);
 
     let parsed_request_message = request_message(&res[..]);
```
it should be possible to write `&buffer[..]` directly

NotBad4U replied:
Sounds good (done in c895390)

Geal commented on src/network/kafka.rs:
```diff
@@ -43,17 +32,18 @@ impl NetworkClient for Client {
+    buffer.put_slice(&mut res[..]);
 
     let parsed_request_message = request_message(&res[..]);
     if let IResult::Done(_, req) = parsed_request_message {
```
we should get the input back and verify that everything was parsed

Geal commented on src/parser/message.rs:
```diff
@@ -57,35 +59,40 @@ pub fn message_set<'a>(input: &'a [u8], size: i32) -> IResult<&'a [u8], MessageS
-        messages
-      })
-  })
+  flat_map!(input, ms_bytes, |msb| toto(msb))
```
why not write `toto` instead of `|msb| toto(msb)`?

NotBad4U replied:
Removed because no we use `messages_set_messages directly`.

Geal commented on src/parser/message.rs:
```diff
@@ -57,35 +59,40 @@ pub fn message_set<'a>(input: &'a [u8], size: i32) -> IResult<&'a [u8], MessageS
+}
+
+fn toto<'a>(msb: &'a [u8]) -> IResult<&'a [u8], MessageSet<'a>> {
+  do_parse!(
```
this could be replaced with `messages_set_messages(msb)`. So `flat_map!(input, ms_bytes, |msb| toto(msb))` could be replaced with `flat_map!(input, ms_bytes, messages_set_messages)`

NotBad4U replied:
done in 7e1ec21

Commit:
Remove the start of the zookeeper handler server

Commit:
Refactor the read_size following the code review

We use a nom macro to read the size from the network

NotBad4U commented:
Thanks for the `Slab` tips. I'll switch to it when the project will be stable again.

Commit:
Remove toto parser and use message_set_message instead following the review

Commit:
Re add the beautifull signature following the review

Commit:
Pass &buffer[..] directly to request_message in kafka handler

Commit:
Remove mmap dep and switch to memmap

This library seem to be the more active on crates.io

Commit:
Refactore storage with using the new lib memmap

Commit:
Loop in write message to be sure to send the complete message

NotBad4U committed:
Ignore mmap test file create after each cargo run

NotBad4U committed:
Add new Client state: Connection

The new KafkaClient connects directly to Kafka brokers instead of connecting to zookeeper for broker discovery. But he need to request all the metadata before sending any data

NotBad4U committed:
Fix the action execute in order of the readiness of an socket

In the first version if the socket was Readable

NotBad4U committed:
Fix warnings: remove unused code

NotBad4U committed:
Upgrade kafka-node client version (test purpose, we need the new KafkaClient)

NotBad4U committed:
Refactore the nodejs kafka client by using the new Client: KafkaClient

NotBad4U committed:
Remove zookeeper part

NotBad4U committed:
Remove connect client state and fix client_read

NotBad4U committed:
Set the port to 9092 (default kafka port)
